pkg/engine/executor/evaluator.go

```diff
@@ -0,0 +1,111 @@
+package executor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+
+	"github.com/apache/arrow-go/v18/arrow/memory"
+	"github.com/grafana/loki/v3/pkg/engine/planner/physical"
+)
+
+// TODO(rfratto): document constraints on how to properly write process
+// methods; particularly around guaranteeing that no Record leaks:
+//
+// * Each process method iterator must yield an error OR generate and yield at
+//   most one Record. Iterators may not generate multiple Records in advance.
+//
+// * Each iterator consumer must Release() the Record after it is done with it.
+//
+// It is safe for iterators to stop iterating early if yield returns false, as
+// future Records won't have been generated yet.
+
+type evaluator struct {
+	opts Options
+	plan *physical.Plan
+	mem  memory.Allocator
+}
+
+func newEvaluator(opts Options, plan *physical.Plan) *evaluator {
+	return &evaluator{opts: opts, plan: plan}
+}
+
+func (e *evaluator) processNode(ctx context.Context, n physical.Node) iter.Seq[Result] {
+	children := e.plan.Children(n)
+
+	iters := make([]iter.Seq[Result], len(children))
+	for i, child := range children {
+		iters[i] = e.processNode(ctx, child)
+	}
+
+	return func(yield func(Result) bool) {
+		switch n := n.(type) {
+		case *physical.DataObjScan:
+			e.processDataObjScan(n, iters)(yield)
+		case *physical.Limit:
+			e.processLimit(n, iters)(yield)
+		default:
```

Comment:
How do you feel about using a visitor pattern like in [logsl/syntax/visit.go](https://github.com/grafana/loki/blob/main/pkg/logql/syntax/visit.go) instead?

Originally I thought pattern matching would do as with the walk implementation. However, there were a few issues with that approach.

1. The Go compiler does not check for exhaustiveness in pattern matches. This lead to bugs.
2. The walk order would be somewhat random.
3. The visitor would give a structure for keeping intermediate results.

I have to admit that I'm not totally convinced the visitor is a great choice especially when one is only interested in specific node types. It did work well for the clone method, though.

Reply:
It just crossed my mind. How do you feel about a stack based iteration to remove the recursion?